File: src/main/java/wraith/fabricaeexnihilo/modules/fluids/MilkFluid.java
package wraith.fabricaeexnihilo.modules.fluids;

import net.fabricmc.fabric.api.transfer.v1.fluid.FluidConstants;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidStorage;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
import net.fabricmc.fabric.api.transfer.v1.fluid.base.FullItemFluidStorage;
import net.minecraft.block.FluidBlock;
import net.minecraft.fluid.Fluid;
import net.minecraft.item.Item;
import net.minecraft.item.Items;
import wraith.fabricaeexnihilo.modules.ModFluids;
import wraith.fabricaeexnihilo.modules.base.AbstractFluid;
import wraith.fabricaeexnihilo.modules.base.FluidSettings;

@SuppressWarnings("UnstableApiUsage")
public class MilkFluid extends AbstractFluid {
    private static final FluidSettings FLUID_SETTINGS = new FluidSettings("milk", 0xFFFFFF, false, false);
    public static final Item BUCKET = Items.MILK_BUCKET;
    public static final MilkFluid FLOWING = new MilkFluid(false);
    public static final MilkFluid STILL = new MilkFluid(true);
    public static final FluidBlock BLOCK = new FluidBlock(STILL, ModFluids.getBlockSettings());

    static {
        // Milk buckets
//        FluidStorage.ITEM.registerForItems((stack, context) -> new FullItemFluidStorage(context, Items.BUCKET, FluidVariant.of(MilkFluid.STILL), FluidConstants.BUCKET), Items.MILK_BUCKET);
        FluidStorage.combinedItemApiProvider(BUCKET).register((context) -> new FullItemFluidStorage(context, Items.BUCKET, FluidVariant.of(MilkFluid.STILL), FluidConstants.BUCKET));
    }

    public MilkFluid(boolean isStill) {
        super(isStill, FLUID_SETTINGS,
                () -> BLOCK,
                () -> Items.MILK_BUCKET,
                () -> FLOWING,
                () -> STILL
        );
    }

    @Override
    public boolean matchesType(Fluid fluid) {
        return fluid == STILL || fluid == FLOWING;
    }

}

File: src/main/kotlin/exnihilofabrico/modules/fluids/MilkFluid.kt
package exnihilofabrico.modules.fluids

import exnihilofabrico.id
import exnihilofabrico.modules.ModFluids
import exnihilofabrico.modules.base.AbstractFluid
import exnihilofabrico.modules.base.FluidSettings
import net.minecraft.fluid.Fluid
import net.minecraft.item.Items
import java.util.function.Supplier

class MilkFluid(isStill: Boolean): AbstractFluid(isStill, fluidSettings, Supplier { Block }, Supplier { Bucket }, Supplier { Flowing }, Supplier { Still }) {
    override fun matchesType(fluid: Fluid) = fluid == Still || fluid == Flowing
    companion object {
        val fluidSettings = FluidSettings(id("block/milk_flow"), id("block/milk_still"), false)
        val Still = MilkFluid(true)
        val Flowing = MilkFluid(false)
        val Bucket = Items.MILK_BUCKET
        val Block = MilkBlock(Still, ModFluids.blockSettings)
    }
}